File: plugins/OrchidJavadoc/src/main/java/com/eden/orchid/javadoc/menu/AllPackagesMenuItemType.java
package com.eden.orchid.javadoc.menu;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItem;
import com.eden.orchid.api.theme.menus.menuItem.OrchidMenuItemImpl;
import com.eden.orchid.api.theme.pages.OrchidPage;
import com.eden.orchid.javadoc.JavadocModel;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class AllPackagesMenuItemType extends OrchidMenuItem {

    private final JavadocModel model;

    @Inject
    public AllPackagesMenuItemType(OrchidContext context, JavadocModel model) {
        super(context, "javadocPackages", 100);
        this.model = model;
    }

    @Override
    public List<OrchidMenuItemImpl> getMenuItems() {
        List<OrchidMenuItemImpl> items = new ArrayList<>();
        if (model.getAllPackages() != null) {
            List<OrchidPage> pages = new ArrayList<>(model.getAllPackages());
            pages.sort(Comparator.comparing(OrchidPage::getTitle));
            items.add(new OrchidMenuItemImpl(context, "All Packages", pages));
        }
        return items;
    }
}

File: plugins/OrchidJavadoc/src/main/kotlin/com/eden/orchid/javadoc/menu/AllPackagesMenuItemType.kt
package com.eden.orchid.javadoc.menu

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.options.annotations.StringDefault
import com.eden.orchid.api.theme.menus.MenuItem
import com.eden.orchid.api.theme.menus.OrchidMenuFactory
import com.eden.orchid.api.theme.pages.OrchidPage
import com.eden.orchid.javadoc.models.JavadocModel
import java.util.ArrayList

@Description("All Javadoc package pages.", name = "Javadoc Packages")
class AllPackagesMenuItemType : OrchidMenuFactory("javadocPackages") {

    @Option
    @StringDefault("All Packages")
    lateinit var title: String

    override fun getMenuItems(context: OrchidContext): List<MenuItem> {
        val model = context.resolve(JavadocModel::class.java)

        val items = ArrayList<MenuItem>()
        val pages = ArrayList<OrchidPage>(model.allPackages)
        pages.sortBy { it.title }
        items.add(
            MenuItem.Builder(context)
                .title(title)
                .pages(pages)
                .build()
        )
        return items
    }

}
